js/userDatabase.js

```diff
@@ -0,0 +1,118 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+(function(exports) {
+  const DATABASE_VERSION = 1;
+  const DATABASE_NAME = "NewTabData";
+
+  const OBJECT_STORE_PREFS = "prefs";
+  const PINNED_LINKS_PREF = "pinnedLinks";
+
+  const gUserDatabase = {
+    _database: null,
+
+    init(mockDB) {
+      return new Promise((resolve, reject) => {
+        var db = mockDB || window.indexedDB;
+        var request = db.open(DATABASE_NAME, DATABASE_VERSION);
+
+        request.onerror = event => {
+          gUserDatabase._logError(event, "Cannot open an indexedDB connection");
+          reject(new Error(event.target.errorCode));
+        };
+        request.onsuccess = event => {
+          gUserDatabase._onDatabaseOpenSuccess(event).then(resolve);
+        };
+        request.onupgradeneeded = event => {
+          gUserDatabase._onDatabaseUpgrade(event);
+        };
+      });
+    },
+
+    save(objectStoreToWrite, objectStoreType, data, mockObjectStore) {
+      return new Promise((resolve, reject) => {
+        var transaction = gUserDatabase._database.transaction([objectStoreToWrite], "readwrite");
+        var objectStore = mockObjectStore || transaction.objectStore(objectStoreToWrite);
+
+        var request = objectStore.get(objectStoreType);
+        request.onsuccess = () => {
+          gUserDatabase._onWriteFetchRequestSuccess(request, data, objectStore, objectStoreType).then(resolve, reject);
+        };
+        request.onerror = event => {
+          gUserDatabase._logError(event, "Failed to store object of type " + objectStoreType);
+          reject(new Error(event.target.errorCode));
+        };
+      });
+    },
+
+    load(objectStoreToRead, objectStoreType) {
+      return new Promise((resolve, reject) => {
+        var transaction = gUserDatabase._database.transaction([objectStoreToRead]);
+        var objectStore = transaction.objectStore(objectStoreToRead);
+        var request = objectStore.get(objectStoreType);
+        var transactionDescription = "Load data " + objectStoreType;
+        gUserDatabase._setSimpleRequestHandlers(request, transactionDescription).then(resolve, reject);
+      });
+    },
+
+    _logError(event, errorString) {
```

Comment:
The problem with this function is that it doesn't show us the context of the error (i.e., where the error actually happened. I would suggest dropping _logError and just console.log() the error in place).